tket/src/Mapping/MappingManager.cpp

```diff
@@ -0,0 +1,111 @@
+// limitations under the License.
+
+#include "Mapping/MappingManager.hpp"
+
+#include "OpType/OpTypeFunctions.hpp"
+#include "TokenSwapping/main_entry_functions.hpp"
+
+namespace tket {
+
+MappingManager::MappingManager(const ArchitecturePtr& _architecture)
+    : architecture_(_architecture) {}
+
+bool MappingManager::route_circuit(
+    Circuit& circuit,
+    const std::vector<RoutingMethodPtr>& routing_methods) const {
+  return this->route_circuit_with_maps(
+      circuit, routing_methods, std::make_shared<unit_bimaps_t>());
+}
+
+bool MappingManager::route_circuit_with_maps(
+    Circuit& circuit, const std::vector<RoutingMethodPtr>& routing_methods,
+    std::shared_ptr<unit_bimaps_t> maps) const {
+  // Assumption; Routing can not route a circuit
+  // with more logical qubits than an Architecture has
+  // physical qubits physically permitted
+  if (circuit.n_qubits() > this->architecture_->n_nodes()) {
+    std::string error_string =
+        "Circuit has" + std::to_string(circuit.n_qubits()) +
+        " logical qubits. Architecture has " +
+        std::to_string(this->architecture_->n_nodes()) +
+        " physical qubits. Circuit to be routed can not have more "
+        "qubits than the Architecture.";
+    throw MappingManagerError(error_string);
+  }
+
+  // mapping_frontier tracks boundary between routed & un-routed in circuit
+  // when initialised, boundary is over output edges of input vertices
+  std::shared_ptr<MappingFrontier> mapping_frontier;
+  if (maps) {
+    mapping_frontier = std::make_shared<MappingFrontier>(circuit, maps);
+  } else {
+    mapping_frontier = std::make_shared<MappingFrontier>(circuit);
+  }
+  // updates routed/un-routed boundary
+
+  mapping_frontier->advance_frontier_boundary(this->architecture_);
+
+  auto check_finish = [&mapping_frontier]() {
+    for (const std::pair<UnitID, VertPort>& pair :
+         mapping_frontier->quantum_boundary->get<TagKey>()) {
+      Edge e = mapping_frontier->circuit_.get_nth_out_edge(
+          pair.second.first, pair.second.second);
+      Vertex v = mapping_frontier->circuit_.target(e);
+
+      if (!is_final_q_type(
+              mapping_frontier->circuit_.get_OpType_from_Vertex(v))) {
+        return false;
+      }
+    }
+    return true;
+  };
+
+  bool circuit_modified = !check_finish();
+  while (!check_finish()) {
+    // The order methods are passed in std::vector<RoutingMethod> is
+    // the order they are iterated through to call "check_method"
+    // If a method performs better but only on specific subcircuits,
+    // rank it earlier in the passed vector
+    bool valid_methods = false;
+    for (const auto& rm : routing_methods) {
+      // true => can use held routing method
+      if (rm->check_method(mapping_frontier, this->architecture_)) {
+        valid_methods = true;
+        unit_map_t partial_permutation =
+            rm->routing_method(mapping_frontier, this->architecture_);
```

Comment:
I do think this would be more efficient. My intention had been to make this change in a different PR after this one as I suspect it will be a bit of an involved task and I don't want it acting as a blocker here. However as I guess it would act as an API change would you prefer we made the changes for this PR?

Reply:
If it's just an API change in the C++, fine to do it as a follow-up task.
If it's a _breaking_ API change in pytket (i.e. would invalidate python code that works with this version), then we should probably do it now.

Reply:
As discussed, we can merge this PR first but should make this change before pytket 1.0. Could you make a Jira task in the "pytket 1.0" story for this and add it to the next sprint?